lang/c/src/value-read-resolution.c

```diff
@@ -0,0 +1,628 @@
+            if(avro_schema_match(wschema, branch_schema))
+            {
+                avro_value_t  branch_dest;
+                avro_value_set_branch(dest, i, &branch_dest);
+                return read_value_with_resolution(reader, source, &branch_dest);
+            }
+        }
+        avro_set_error("There is no compatible branch in readers schema!");
+        return EINVAL;
+    }
+    else if ((rtype != AVRO_UNION) && (wtype == AVRO_UNION))
+    {
+        // writer is union, reader is not
+        // must read discriminant value (long) from buffer
+        // before reading actual branch value.
+
+        int rval;
+        int64_t discriminant;
+        int64_t  branch_count;
+        avro_value_t  branch;
+
+        check_prefix(rval, avro_binary_encoding.
+                     read_long(reader, &discriminant),
+                     "Cannot read union discriminant: ");
+
+        branch_count = avro_schema_union_size(wschema);
+
+        if (discriminant < 0 || discriminant >= branch_count) {
+            avro_set_error("Invalid union discriminant value: (%d)",
+                           discriminant);
+            return EINVAL;
+        }
+
+        check(rval, avro_value_set_branch(source, discriminant, &branch));
+        check(rval, avro_value_get_current_branch(source, &branch));
+        return read_value_with_resolution(reader, &branch, dest);
+    }
+    else
+    {
+        // Shouldn't be here
+        avro_set_error("Cannot resolve unions.");
+        return EINVAL;
+    }
+}
+
+/*
+ * Read of maps, arrays and unions is mostly the same as when resolution
+ * is not needed, the only thing is that we need to keep track of `source`
+ * value to be able to resolve schemas recursively.
+ */
+int resolve_unions(avro_reader_t reader, avro_value_t *source, avro_value_t *dest)
+{
+    int rval;
+    int64_t discriminant;
+    avro_schema_t  union_schema_source;
+    int64_t  branch_count_source;
+    avro_value_t  branch_source;
+    
+    avro_schema_t  union_schema_dest;
+    int64_t  branch_count_dest;
+    avro_value_t  branch_dest;
+    
+    check_prefix(rval, avro_binary_encoding.
+                 read_long(reader, &discriminant),
+                 "Cannot read union discriminant: ");
+
+    union_schema_source = avro_value_get_schema(source);
+    branch_count_source = avro_schema_union_size(union_schema_source);
+    
+    union_schema_dest = avro_value_get_schema(dest);
+    branch_count_dest = avro_schema_union_size(union_schema_dest);
+    if (discriminant < 0 || discriminant >= branch_count_source) {
+        avro_set_error("Invalid union discriminant value: (%d)",
+                       discriminant);
+        return EINVAL;
+    }
+    
+    check_prefix(rval,
+                 avro_value_set_branch(source, discriminant, &branch_source),
+                 "Cannot set current branch");
+    avro_schema_t source_branch_schema = avro_value_get_schema(&branch_source);
+    int i = 0;
```

Comment:
do you need this?

Reply:
yes, C99 doesn't allow `for(int i = 0 ...)`